### ISSUE TITLE
Setting a password on account sign up with expired token gives wrong error message

### ISSUE THREAD

**Issue by anthony-hull** [Expensify/Expensify.cash#3189]
## Expected Result:
Error to say:
Your token has expired, click here to resend one.

## Actual Result:
Error was as follows:
![image](https://user-images.githubusercontent.com/47184118/119882160-02b4d500-bf26-11eb-878a-e910d9c7f6ba.png)
The user isn't trying to log in at this point.

## Action Performed:
trigger password reset
make sure you're logged out
navigate to here: http://expensify.cash/setpassword/[USERID]/[RESETCODE]
make sure you have credentials.login key set to your email address in local storage
submit a valid password

## Platform:
Where is this issue occurring?

Web

**Version Number:**
**Logs:**
{"jsonCode":407,"title":"Session expired","message":"Your session has expired. Please sign in again.","requestID":"0a50c9906d0000f3ff2984d000000001"}

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @lschurr (`AutoAssignerTriage`), see https://stackoverflow.com/c/expensify/questions/4749 for more details.

**Comment by anthony-hull** [comment]
This looks like it has been changed to something more useful in the backend:
![image](https://user-images.githubusercontent.com/47184118/120825318-2d6be280-c551-11eb-98cc-6d359c0487ec.png)
Possibly still some call to action to resend a code for the user would be good?

**Comment by lschurr** [comment]
Hm, when I tried to reproduce this, I got a privacy error:

![image](https://user-images.githubusercontent.com/33293730/121077065-00574400-c78c-11eb-9209-ebb87e3d8e76.png)

I'm going to add the eng label to have someone else take a look.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @stitesExpensify (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by stitesExpensify** [comment]
Hi @anthony-hull, can you clarify your solution please?

**Comment by anthony-hull** [comment]
I think there should be some way of the user requesting a new set password email link. So maybe replace the whole form with a request new link button?

**Comment by anthony-hull** [comment]
It looks like the API now re-sends en email when you use an expired token. But I've been using the reset password UX rather than set password for the a new user for most of my testing.

**Comment by stitesExpensify** [comment]
I think it makes sense to replace the form with a new link button in this case. Are you still interested in this @anthony-hull ?

**Comment by stitesExpensify** [comment]
Bump @anthony-hull 😄

**Comment by anthony-hull** [comment]
I've just checked this flow and the API now just resends a new email if if gets a bad magic code. The button wouldn't be needed. 

But informing the user that a new email has been sent would be an improvement I think.

**Comment by stitesExpensify** [comment]
Cool, I'm in favor of you making that solution as those seem like great improvements to me! If you want to update the OP I'll add the external label and we can get you working on it! We are still currently in a merge hold though so I'm not sure when it will get merged/paid out FYI

**Comment by anthony-hull** [comment]
ok :)

**Comment by stitesExpensify** [comment]
Looks good! I updated the copy a little bit and we should be good to go

**Comment by anthony-hull** [comment]
I like the tweak. Does this need an external label now?

**Comment by mallenexpensify** [comment]
@anthony-hull , can you provide an update on when we can expect a PR? Thanks

**Comment by anthony-hull** [comment]
Hey @mallenexpensify I'm working on this today. 

Apologies, I incorrectly assumed that the API would behave differently. 

I've found out that you don't get a new code sent out if you mutilate the token if your account is still awaiting validation. But if you have already validated your account and you use that page with a malformed code it will resend one.

The above steps and discussion were based on this assumption as it saved me creating multiple new user accounts, so the majority of my testing was against a user that was already validated. Now I can't get a fast feedback without waiting a few days for the code to expire.

When mutilating the magic code in the URL sent during a new user sign up and submitting the form, a new email is not sent. This changes the UX a little. 

The user can only get to this point if they some how copy pasted the URL from the email wrong or edited it in the URL bar. The only way for the user to get out of this situation is to be told to click the link in their emails again. There won't be a new one waiting like the proposed copy changes outline. **Do we even want to cover this situation?**

I'm now not 100% sure if a new email is sent when an expired token is used by an unverified user. I made this issue in May so I can't say with confidence if I tested this! I hope it does else it is a dead end for the user.
I will have to wait at least 2 days to test this myself as I need to wait for the code to expire. (password reset links from expensify.com expire after 48 hours, Marc seems to think it might be 7 days for this link). I'm on holiday from 5-9th Nov so if we need to wait for me to test this before being able to make a decision I won't be able to proceed on this tomorrow and will return to it on the 9th


Proposed solutions:

- Make the API always send a new email link out when an invalid code is used by an account which hasn't been validated, like it does when using a token made by the reset password flow from the main site in the testing steps above and continue this issue as previously agreed.

- Only handle the expired code case and ignore the user somehow causing the code to be wrong and continue this issue as previously agreed.

- Handle the malformed code and expired code cases differently. This would require the API to distinguish these error states in the response. Right now I can only see the response from sending a malformed code (I'm not sure how long the codes are valid, so I can't confirm if the response changes for expired codes):
```
{
    "code": 666,
    "jsonCode": 666,
    "type": "Expensify\\Libs\\Error\\ExpError",
    "UUID": "42742477-542E-4EA8-876A-7842D0FF2C43",
    "message": "Email validation failed",
    "title": "",
    "data": [],
    "htmlMessage": "",
    "requestID": "6a875270fe7c0672-SAN"
}
```

**Comment by mallenexpensify** [comment]
> The user can only get to this point if they some how copy pasted the URL from the email wrong or edited it in the URL bar.
Do we even want to cover this situation?

I think it's an edge case that we likely don't need to focus on. 

> Make the API always send a new email link out when an invalid code is used by an account which hasn't been validated

Not an engineer but.... this makes sense to me.  ie. Don't make the user click another button if we already know their token is expired and that they need a new email sent to them. 

>Only handle the expired code case and ignore the user somehow causing the code to be wrong and continue this issue as previously agreed.

Also agree with this, I'm fine with _nothing_ happening if a malformed token/link is used. Maybe we just send them back to the main sign-in page?

**Comment by anthony-hull** [comment]
Thanks Matt. I'm waiting on my current code to expire before I can proceed. I need to see the API response when I hit it with an expired code and I don't know of a way to force this without access to the backend/database. 

I will be able to better update you and progress the PR once I have that information. 
I'll update you again on the 10th November when I'm back from holiday and can test this.

**Comment by anthony-hull** [comment]
I just tested a link after waiting 7 days and it worked fine. 
Maybe it never expires. 
If so this issue is moot unless you intend for the link to expire and this is a back end bug/enhancement?

**Comment by mallenexpensify** [comment]
> Maybe it never expires.
@anthony-hull , would you mind raising this in #expensify-open-source to get more 👀? I'm not sure if we want it to expire or not.

**Comment by anthony-hull** [comment]
the link didn't work after the weekend, so 11 days since. Possibly it's some time after 7 days. 

But I have a new issue. It appears I don't get a new link send. I've started a thread on it:
https://expensify.slack.com/archives/C01GTK53T8Q/p1636974122197400

**Comment by michaelhaxhiu** [comment]
Just to check, this is still blocked @anthony-hull?

**Comment by anthony-hull** [comment]
yes it is. But I can do the work and get the PR reviewed. but it only makes sense to merge once the backend is fixed.

**Comment by michaelhaxhiu** [comment]
>But I can do the work and get the PR reviewed. but it only makes sense to merge once the backend is fixed.

If it's blocked on merging until that issue is fixed, then feel free to continue prioritizing how you have been thus far!

>The linked issue is above. Is there any conversation internally about it?

I see no conversation on the issue yet, latest progress should always been visible on the issue itself.

**Comment by anthony-hull** [comment]
I will unlikely be able to work on this further until I return on the 14th Dec

**Comment by michaelhaxhiu** [comment]
@anthony-hull thanks for the heads up! Feel free to resume on this when you return.

**Comment by michaelhaxhiu** [comment]
@anthony-hull was offline last week for vacation, and returned this week. Blocking issue was fixed and so Anthony should be unblocked to resume progress on his PR https://github.com/Expensify/App/pull/6587

**Comment by anthony-hull** [comment]
I'm working on this issue next :) Just have a small one to close out first.

**Comment by michaelhaxhiu** [comment]
Awesome thanks for the update!

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kadiealexander (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by michaelhaxhiu** [comment]
I'm re-applying the `External` label on this GH to trigger auto-assignment to a fellow CM, as I'm going OOO for the next 2 weeks.

**Comment by anthony-hull** [comment]
Have a great holiday @michaelhaxhiu!

**Comment by MelvinBot** [comment]
Triggered auto assignment to @SofiedeVreese (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by kadiealexander** [comment]
Reassigning as I'll be ooo until January 10th, thanks @SofiedeVreese!!

**Comment by SofiedeVreese** [comment]
Omg lol same here. I'm also going to be ooo, until January 11th welp.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @kadiealexander (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by mallenexpensify** [comment]
Apologies for the label/assigning spam, wanted to check to see who it got assigned to.  @kadiealexander and @SofiedeVreese , I'm around through the holidays so I'll keep tabs on this.

**Comment by kadiealexander** [comment]
You're a champion Matt, thank you!!

**Comment by mallenexpensify** [comment]
@anthony-hull do you have an update on progress?  We recently [updated our CONTRIBUTING.md ](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#:~:text=Daily%20updates%20on,original%20contract%20terminated.)to include details on expected timeliness of working on issue . 

> Daily updates on weekdays are highly recommended. If you know you won’t be able to provide updates for > 1 week, please comment on the PR or issue how long you plan to be out so that we may plan accordingly. We understand everyone needs a little vacation here and there. Any issue that doesn't receive an update for 1 full week may be considered abandoned and the original contract terminated.

Just wanna make sure the PR will be submitted soon.

**Comment by anthony-hull** [comment]
thanks!
I'm working on this tonight and tomorrow, I updated the linked PR earlier this evening. I hope to have it ready for review tomorrow

**Comment by anthony-hull** [comment]
I haven't managed to finish the PR today. 
I don't foresee any blockers for getting it finished tomorrow. 
Currently I'm working through the edge cases listed in my tests.

**Comment by mallenexpensify** [comment]
Thanks for the update @anthony-hull !

**Comment by anthony-hull** [comment]
I encountered extra complexity. I will carry on working on this tomorrow
I also have an outstanding issue with another issue that this PR will also close which needs to be answered. I can carry on with the rest of the functionality without this input however

**Comment by anthony-hull** [comment]
I resolved the above issue. But I have a new outstanding question in the PR.

**Comment by anthony-hull** [comment]
still pending input on question

**Comment by mallenexpensify** [comment]
Appreciate the updates @anthony-hull !

**Comment by anthony-hull** [comment]
The PR linked is ready for review, I have requested for such. Just missing some translations and response on required testing documentation!
Hopefully get this one merged soon!

**Comment by anthony-hull** [comment]
Waiting on PR review, answer to a question and a backend bug to be confirmed reproduced by the reviewer and fixed

**Comment by mallenexpensify** [comment]
looks like PR is on staging! https://github.com/Expensify/App/pull/6587

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.33-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/6587


If no regressions arise, payment will be issued on 2022-02-08. :confetti_ball:

**Comment by mallenexpensify** [comment]
Thanks @anthony-hull for the help and persistence with this issue. Paid in Upwork, $500 - $250ea for reporting and the fix
